pkg/trie/triedb/node.go

```diff
@@ -168,8 +207,204 @@ func newNodeFromEncoded(nodeHash common.Hash, data []byte, storage NodeStorage)
 			children[i] = child
 		}
 
-		return Branch{partialKey: key, children: children, value: NewFromEncoded(value)}, nil
+		return Branch{partialKey: key, children: children, value: NewValueFromEncoded(encoded.PartialKey, value)}, nil
+	default:
+		panic("unreachable")
+	}
+}
+
+type NodeToEncode interface {
+	isNodeToEncode()
+}
+
+type (
+	NewNodeToEncode struct {
+		partialKey []byte
+		value      []byte
+	}
+	TrieNodeToEncode struct {
+		child NodeHandle
+	}
+)
+
+func (NewNodeToEncode) isNodeToEncode()  {}
+func (TrieNodeToEncode) isNodeToEncode() {}
+
+type ChildReference interface {
+	isChildReference()
+	getNodeData() []byte
+}
+
+type (
+	HashChildReference struct {
+		hash common.Hash
+	}
+	InlineChildReference struct {
+		encodedNode []byte
+	}
+)
+
+func (HashChildReference) isChildReference() {}
+func (h HashChildReference) getNodeData() []byte {
+	return h.hash.ToBytes()
+}
+func (InlineChildReference) isChildReference() {}
+func (i InlineChildReference) getNodeData() []byte {
+	return i.encodedNode
+}
+
+type childFunc = func(node NodeToEncode, oslice []byte, oindex *byte) (ChildReference, error)
+
+const firstPrefix = (0x00 << 6)
+const emptyTrieBytes = firstPrefix | (0x00 << 4)
+
+// TODO: move this to codec package
+func NewEncodedNode(node Node, childF childFunc) (encodedNode []byte, err error) {
+	encodingBuffer := bytes.NewBuffer(nil)
+
+	switch n := node.(type) {
+	case Empty:
+		return []byte{emptyTrieBytes}, nil
+	case Leaf:
+		pr := n.partialKey
+		value, err := NewEncodedValue(n.value, pr, childF)
+		if err != nil {
+			return nil, err
+		}
+
+		err = NewEncodedLeaf(pr, value, encodingBuffer)
+		if err != nil {
+			return nil, err
+		}
+	case Branch:
+		var value codec.EncodedValue
+		if n.value != nil {
+			pr := n.partialKey
+			value, err = NewEncodedValue(n.value, pr, childF)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		var children [codec.ChildrenCapacity]ChildReference
+		for i, child := range n.children {
+			if child == nil {
+				continue
+			}
+
+			pr := n.partialKey[len(n.partialKey)-1:] // TODO: check this
+			oindex := byte(i)
```

Comment:
nit/what is oindex? maybe a more clear name could be used

Reply:
Renamed